### ISSUE TITLE
Bug: [no-unnecessary-condition] false-positive for array index access optional chaining on array union type

### ISSUE THREAD

**Issue by JH1ller** [typescript-eslint/typescript-eslint#8540]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=FAFwngDgpgBAgiKBnEMC8MDeMCGAuGFAJwEsA7AcxgF8BuUSWAIURXS1wOPIoBoYARlxClKNesADGAezJsArmRKyAKshAEE6gNoBdGAB8YLHfox6JMuaiLJ5AG1QZFysmpTaADLoD8AOhx6IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=&tokens=false

### Repro Code

```TypeScript
type Atest = { a: string };
type Btest = { a: string, b: string };

const unionTest: Atest[] | Btest[] = [];

const result = unionTest[0]?.a;

// unionTest[0] can be undefined, yet typescript-eslint reports the ?. as unnecessary.
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### Expected Result

Should not report an `Unnecessary optional chain on a non-nullish value` when using optional chaining after an array index access.

### Actual Result

An optional chaining usage after array index access is reported as unnecessary, but only if the value is an array union type.

**Comment by auvred** [comment]
Duplicate of #8281, #8272, #7505, #7118, #7086, #5225, #2883

Explanation from https://github.com/typescript-eslint/typescript-eslint/issues/8272#issuecomment-1900483208:

> See the [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) compiler option.
> 
> The rule is powered by types. Withe the option turned off - the types say the variable `item` is of type `T`. So the rule reports. With the option turned on - the types would say the variable `item` is of type `T | undefined`. So the rule would not report.

**Comment by auvred** [comment]
Looks like a bug! I think we should iterate over union parts in these functions:

https://github.com/typescript-eslint/typescript-eslint/blob/8431d9d396ee9b2a09e5d30f3645c2b538a03dcd/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L176-L183
